packages/jupyterlab-manager/src/manager.ts

```diff
@@ -213,14 +332,114 @@ class WidgetManager extends ManagerBase<Widget> implements IDisposable {
     return this._rendermime;
   }
 
+  /**
+   * A signal emitted when state is restored to the widget manager.
+   *
+   * #### Notes
+   * This indicates that previously-unavailable widget models might be available now.
+   */
+  get restored(): ISignal<this, void> {
+    return this._restored;
+  }
+
   register(data: IWidgetRegistryData) {
     this._registry.set(data.name, data.version, data.exports);
   }
 
+  /**
+   * Get a model
+   *
+   * #### Notes
+   * Unlike super.get_model(), this implementation always returns a promise and
+   * never returns undefined. The promise will reject if the model is not found.
+   */
+  async get_model(model_id: string): Promise<WidgetModel> {
+    const modelPromise = super.get_model(model_id);
+    if (modelPromise === undefined) {
+      throw new Error('widget model not found');
```

Comment:
@jasongrout I initially [added some code here](https://github.com/jupyter-widgets/ipywidgets/pull/2265/commits/6078163210f4c579fb69dc1fa8226657dbdfa913#diff-d4bf928d4fe8ff74753016095895f635R309) where if widgets were in the process of being restored from live / state, this would await the completion of the restore. Was this not needed?

Reply:
I simplified this function, and instead made the renderer responsible for rerendering when a restore happened: https://github.com/jupyter-widgets/ipywidgets/pull/2265/files#diff-cc0d118ac726eae45225b560ae664079R42

One of the reasons to do this was to make the restore procedure simpler, since the restore itself can use get_model to rehydrate references to other models.

Reply:
After talking directly, we determined that the only dangling references for `get_model` would be from trying to render yet-unrestored models. The logic I added was therefore not needed (and introduced a possible deadlock). The current logic deals with such dangling references in the view.